### PR TITLE
Limit the amount of peers on the server

### PR DESCRIPTION
**Pull request by daniel-jodlos**
- Limit the amount of peers on the server
- Limit the amount of peers on sandboxes to 4
- Pass variable through docker-compose.yml

### CODE REVIEW

File: assets/src/pages/room/hooks/useMembraneClient.tsx [hunk 1]
```diff
@@ -135,17 +142,15 @@ export const useMembraneClient = (
         webrtc.join(peerMetadata);
         setWebrtc(webrtc);
       })
-      .receive("error", (_response) => {
-        handleError(`Couldn't establish signaling connection`);
+      .receive("error", (response) => {
+        if (response === "server_full") {
+          handleError("Server has reached maximum allowed load. Disconnecting");
```
Comment (daniel-jodlos) [inline assets/src/pages/room/hooks/useMembraneClient.tsx:147]: The front-end team might want to consider implementing a different way of handling this error, this definitely isn't ideal

**Comment by mickel8** [comment]
Given that our machines got updated, we need to find out new constraints.

I also wonder about room limits because e.g. 4 peers in one room differ from 2 rooms, 2 peers each

**Comment by daniel-jodlos** [comment]
> I also wonder about room limits because e.g. 4 peers in one room differ from 2 rooms, 2 peers each

One big room is the worst-case scenario. I intentionally selected the simplest solution, as we are in a hurry.

> Given that our machines got updated, we need to find out new constraints.

Should be straightforward, if old machines could handle 5 peers, then the new ones should be able to handle 7. But we may in fact want to find these values empirically.

File: assets/src/pages/room/hooks/useMembraneClient.tsx (hunk 1 above)
Comment (bblaszkow06) [reply assets/src/pages/room/hooks/useMembraneClient.tsx:147]: cc: @kamil-stasiak

File: lib/videoroom/peers_tracker.ex [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  @type peer_id() :: Membrane.RTC.Engine.Endpoint.id()
+
+  @spec child_spec([]) :: Supervisor.child_spec()
+  def child_spec([]) do
+    Registry.child_spec(name: __MODULE__, keys: :unique)
+  end
+
+  @spec try_add(peer_id()) :: :ok | {:error, :full}
+  def try_add(peer_id) do
+    if Registry.count(__MODULE__) == Application.fetch_env!(:membrane_videoroom_demo, :max_peers) do
```
Comment (bblaszkow06) [inline lib/videoroom/peers_tracker.ex:16]: I'd suggest rather using `>=`
```suggestion
    if Registry.count(__MODULE__) >= Application.fetch_env!(:membrane_videoroom_demo, :max_peers) do
```

File: lib/videoroom/peers_tracker.ex [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  # It is crutial to ensure that we don't get random timeouts and other overload problems
+
+  @type peer_id() :: Membrane.RTC.Engine.Endpoint.id()
+
+  @spec child_spec([]) :: Supervisor.child_spec()
+  def child_spec([]) do
+    Registry.child_spec(name: __MODULE__, keys: :unique)
+  end
+
+  @spec try_add(peer_id()) :: :ok | {:error, :full}
```
Comment (bblaszkow06) [inline lib/videoroom/peers_tracker.ex:14]:
```suggestion
  @spec try_add(peer_id()) :: :ok | {:error, :server_full}
```

File: lib/videoroom/peers_tracker.ex [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  def try_add(peer_id) do
+    if Registry.count(__MODULE__) == Application.fetch_env!(:membrane_videoroom_demo, :max_peers) do
+      {:error, :server_full}
+    else
+      Registry.register(__MODULE__, peer_id, nil)
```
Comment (bblaszkow06) [inline lib/videoroom/peers_tracker.ex:19]: We shouldn't ignore the returned value here
```suggestion
      {:ok, _owner} = Registry.register(__MODULE__, peer_id, nil)
```

**Comment by mickel8** [comment]
After recent experiments, I believe that on 8 cores we should be able to handle ~10 peers and stay around 80% CPU usage. Even if we set the limit to 8 peers, prod and staging have 16 cores each so the limit of 5 peers is not accurate

**Review by mickel8 — COMMENTED** [review]
After recent experiments, I believe that on 8 cores we should be able to handle ~10 peers and stay around 80% CPU usage. Even if we set the limit to 8 peers, prod and staging have 16 cores each so the limit of 5 peers is not accurate

**Pull request closed without merging**